### ISSUE TITLE
proposal: `discarded_futures`

### ISSUE THREAD

**Issue by pq** [dart-lang/sdk#58747]
## Description

Don't discard futures in async functions.

## Details

Making asynchronous calls in non-`async` functions is usually the sign of a programming error.  In general these functions should be marked `async` and such futures should likely be awaited (as enforced by `unawaited_futures`).

In case you _really do_ want to discard a Future in a non-`async` function, the recommended way is to use `unawaited` from `dart:async`. The `// ignore` and `// ignore_for_file` comments also work.

## Kind

Error

## Good Examples 

```dart
Future<void> recreateDir(String path) async {
  await deleteDir(path);
  await createDir(path);
}

Future<void> deleteDir(String path) async {}

Future<void> createDir(String path) async {}
```

## Bad Examples

```dart
void recreateDir(String path) {
  deleteDir(path);
  createDir(path);
}

Future<void> deleteDir(String path) async {}

Future<void> createDir(String path) async {}
```

## Discussion

See dart-lang/sdk#57653, dart-lang/sdk#58512 and [`unawaited_futures`](https://dart-lang.github.io/linter/lints/unawaited_futures.html).

/fyi @eernstg @diegovar @dowski @bsutton

### Discussion checklist

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [x] If there's any prior art (e.g., in other linters), please add references here.
- [x] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn’t any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/master/doc/WritingLints.MD
[Effective Dart]: https://dart.dev/guides/language/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by bsutton** [comment]
I have a class FuturebulderEx that takes a future and is intended to be
used in a build method.

So it appears to be a discarded future but it's not.

https://pub.dev/packages/future_builder_ex

**Comment by bsutton** [comment]
Ideally I want to be able to mark the interface on FuturebulderEx as 'using
the future' which would then suppress this lint.

**Comment by FMorschel** [comment]
@pq why is this still open? I see the rule https://dart.dev/tools/linter-rules/discarded_futures is available as of Dart 2.18.

**Comment by pq** [comment]
Yeah. I think it would be reasonable to close this and move ongoing conversation to new issues.

**Comment by FMorschel** [comment]
The most common comment I found regarding this feature is related to this issue: https://github.com/dart-lang/sdk/issues/59504. So if anyone still wants this to be addressed please upvote it.

**Comment by pq** [comment]
Ah, great. I'll close this one and we can follow-up on the listed issues.

Thanks!
